blog(embertimes158): blogpost detecting ember comp (#725)
ember-learn/ember-blog: source/2020-07-31-the-ember-times-issue-158.md

---
title: The Ember Times - Issue No. 158
author: Chris Ng, the crowd
tags: Recent Posts, Newsletter, Ember.js Times, Ember Times, 2020
alias : "blog/2020/07/31-the-ember-times-issue-158.html"
responsive: true
---

<SAYING-HELLO-IN-YOUR-FAVORITE-LANGUAGE> Emberistas! 🐹

> Read the blog post on Detecting Ember.js Components Entering or Leaving the Viewport 🔍,

Reviewer: Maybe shorten the tagline?
```suggestion
Detect when Ember components enter or leave the viewport 🔍,
```

<SOME-INTRO-HERE-TO-KEEP-THEM-SUBSCRIBERS-READING>

READMORE

---

## [Detecting Ember.js Components Entering or Leaving the Viewport 🔍](https://medium.com/@koushikrad/using-an-ember-cli-addon-detecting-ember-js-components-entering-or-leaving-the-viewport-dda5ad9b46bf)

> [Koushik Radhakrishnan (@Koushikrad)](https://github.com/Koushikrad) wrote a blog post on [detecting when Ember components are entering or leaving the viewport](https://twitter.com/koushikrad/status/1287419970121379840). Each of the components in the dashboard product, in the blog post, made an API request. But do we need all widgets to fetch its data at the same time when a user lands on a dashboard page? What if the user’s viewport shows only 5 widgets at a time and not 10?

Reviewer: I think the following changes may help illustrate Koushik's example more clearly.
```suggestion
[Koushik Radhakrishnan (@Koushikrad)](https://github.com/Koushikrad) wrote a blog post on [detecting when Ember components are entering or leaving the viewport](https://twitter.com/koushikrad/status/1287419970121379840).

Consider a dashboard with 10 widget components, each of which makes an API request. When a user lands on the dashboard page, do we need all widgets to fetch their data at the same time? Can the user’s viewport show only 5 widgets at a time instead?
```

Reviewer 2: Adding on top of @ijlee2 's suggestion,

```suggestion
[Koushik Radhakrishnan (@Koushikrad)](https://github.com/Koushikrad) wrote a blog post on [detecting when Ember components are entering or leaving the viewport](https://twitter.com/koushikrad/status/1287419970121379840).

Consider a dashboard with 10 widget components, each of which makes an API request. When a user lands on the dashboard page, do we need all the widgets to fetch their data at the same time, even if the user's viewport shows only 5 widgets at a time?
```

> Using the [ember-in-viewport](https://github.com/DockYard/ember-in-viewport) addon which detects if an Ember component has entered the browser's viewport. By default, this uses the IntersectionObserver API if it detects it the DOM element is in your user's browser – failing which, it falls back to using requestAnimationFrame, then if not available, the Ember run loop and event listeners.

Reviewer: The 2nd sentence feels quite technical. I think I'd try to simplify it to something like below, or skip the sentence altogether to keep this writing at a high level.
```suggestion
A solution is to use the [ember-in-viewport](https://github.com/DockYard/ember-in-viewport) addon, which detects if an Ember component has entered the browser's viewport. The addon tries a few different approaches (`IntersectionObserver` API, then `requestAnimationFrame`, then the Ember run loop and event listeners) to detect if a DOM element is in the user's browser.
```

Reviewer 2: I agree with the same that it feels technical.

> By hooking up with the addon’s provided `inViewport` service we are able to only request data once the component is within the viewport.

Reviewer: ```suggestion
By hooking up with the addon’s provided `inViewport` service, we are able to request data once the component is within the viewport.
```

> ```

Reviewer: Linting asked for specifying the language in a code block.
```suggestion
```javascript
```

const { onEnter } = this.inViewport.watchElement(this.element, { viewportTolerance });  
onEnter(this._renderInView.bind(this));
```

> The addon also provides the `viewportTolerance` property which is a configuration value to signal when a component is considered to be entered or leaving the viewport. The components that are not in view will not make a request for API data, unlike the route’s model hook using a `Promise.all`. Read more on this at the [full blog post](https://medium.com/@koushikrad/using-an-ember-cli-addon-detecting-ember-js-components-entering-or-leaving-the-viewport-dda5ad9b46bf)!

Reviewer: I think I'd skip the 1st sentence for brevity. 🙂
```suggestion
The components that are not in view will not make a request for API data, unlike the route’s model hook using a `Promise.all`.

Read more about detecting components from the [blog post](https://medium.com/@koushikrad/using-an-ember-cli-addon-detecting-ember-js-components-entering-or-leaving-the-viewport-dda5ad9b46bf)!
```
